llama.cpp

```diff
@@ -13044,6 +13044,63 @@ void llama_sample_min_p(struct llama_context * ctx, llama_token_data_array * can
     }
 }
 
+void llama_sample_dry(llama_token_data_array * candidates, const llama_token * last_tokens, size_t last_tokens_size, float dry_base, float dry_multiplier, int dry_allowed_length, const llama_token * seq_breakers, size_t seq_breakers_size) {
+    // loop through each candidate
+    for (size_t i = 0; i < candidates->size; ++i) {
```

Comment:
Unless I'm misunderstanding something, this implementation is highly inefficient.

You're looping through all candidates. Which, before truncation samplers are applied, are simply the full vocabulary of the tokenizer. For today's LLMs, that's typically between 32k and 200k.

Then in the inner loop, you're going over all tokens in the context. Which can be tens of thousands of tokens as well.

So we're talking at minimum `(candidates->size - seq_breakers_size) * last_tokens_size` iterations, which is possibly hundreds of millions. And that's not even counting the innermost sequence matching loop!

My original implementation works like this instead:

1. Find all occurrences of the last token in the context. This is typically a small percentage of the context size.
2. For each occurrence, try to extend the match backwards as far as possible. Which is usually 0-2 tokens.
3. Apply penalties to the tokens following the relevant matches.

In practice, this amounts to a single search over the context, plus a small overhead where matches are found. That's *five orders of magnitude* less work than your current implementation, assuming a vocabulary size in the tens of thousands.

Reply:
Thanks for the suggestions! Your method makes much more sense.

I think I didn't notice this because I was using the default top_k = 40 in llama.cpp, so it was only 40 tokens max 😂

Reply:
DRY should be applied *before* any truncation samplers such as top_k. That's because you want sufficiently penalized tokens to be truncated away by those samplers. Otherwise you can be left with a distribution that violates the coherence properties that truncation is supposed to provide.

So this is another important thing that needs to be changed.

Reply:
@p-e-w I have updated the code with your suggestions.

Looking more closely, I believe I misunderstood some of your logic in my last implementation (which was pretty naive 😂), I've refactored the sampler to better match your original implementation. Please take a look when you get the chance!

Currently, DRY is applied just after the repetition penalty, before any top k and temperature samplers.